### PR TITLE
Add Observer class

### PR DESCRIPTION
**Pull request by Aschen**
## What does this PR do ?

Depends on https://github.com/kuzzleio/sdk-javascript/pull/672

Add an `Observer` class to manipulate `RealtimeDocument`.

Those documents have the same structure as the `Document` object from the Document Controller except that they are connected to the realtime engine and their content will mutate when a realtime notification about changes is received.

```js
const observer = new Observer(sdk);

// SearchResult containing realtime documents
let realtimeResult = await observer.search('nyc-open-data', 'yellow-taxi', { query }, options);

// Dispose ressources (realtime connection + memory)
observer.stop('nyc-open-data', 'yellow-taxi');
```

### CODE REVIEW

File: src/core/Observer.ts [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+      observedDocuments.delete(document._id);
+    }
+
+    return this.resubscribe(index, collection);
+  }
+
+  /**
+   * Unsubscribe from every collections and clear all the realtime documents.
+   */
+  dispose () {
```
Comment (rolljee) [inline src/core/Observer.ts:115]: since there is a stop method, i would call it `stopAll ()` instead wdyt ?

**Review by rolljee — APPROVED** [review]
LGTM, just a nit on naming convention

File: src/core/Observer.ts (hunk 1 above)
Comment (Aschen) [reply src/core/Observer.ts:115]: Actually I was thinking about having only one method that can take index & collection, index & collection & documents or nothing to stop subscriptions.  
I will update and you can tell me

**Review by xbill82 — CHANGES_REQUESTED** [review]
WIP Review

File: doc/7/core-classes/observer/stop/index.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+stop (index?: string, collection?: string, documents?: Array<{ _id: string; }>): Promise<void>
+```
+
+| Argument | Type | Description |
+|----------|------|-------------|
+| `index` | <pre>string</pre> | Index name |
+| `collection` | <pre>string</pre> | Collection name |
+| `documents` | <pre>{ _id: string; }[]</pre> | Array of documents |
+
+## Usage
```
[the remark this answers is not in the document]
Comment (Aschen) [reply doc/7/core-classes/observer/stop/index.md:30]: Fixed :+1:

**Review by rolljee — APPROVED** [review]

**Review by xbill82 — APPROVED** [review]

File: src/core/RealtimeDocument.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import { JSONObject } from '../types';
+
+/**
+ * Represents a realtime document that will mutate it's content in realtime
+ * according to database changes.
+ *
+ * Use the `Observer` class to manipulate such documents.
+ */
+export class RealtimeDocument {
```
[the remark this answers is not in the document]
Comment (Aschen) [reply src/core/RealtimeDocument.ts:9]: There is a lot of work to do on our types yes.. 
And it should be done both for the SDK and Kuzzle

**Pull request merged**